src/foam/core/XMLSupport.java

```diff
@@ -0,0 +1,199 @@
+              return;
+            }
+            break;
+        }
+      }
+    } catch (XMLStreamException ex) {
+      throw new XMLStreamException("Premature end of xml file");
+    }
+  }
+
+  public static void toXML(List<FObject> objList, XMLStreamWriter xmlStreamWriter) {
+    try {
+      xmlStreamWriter = new IndentingXMLStreamWriter(xmlStreamWriter);
+      Iterator i = objList.iterator();
+      xmlStreamWriter.writeStartDocument();
+      //Root tag
+      xmlStreamWriter.writeStartElement("objects");
+
+      while ( i.hasNext() ) {
+        toXML((FObject) i.next(), xmlStreamWriter);
+      }
+
+      xmlStreamWriter.writeEndElement();
+      xmlStreamWriter.writeEndDocument();
+
+    } catch (XMLStreamException ex) {
+    }
+  }
+
+  public static void toXML(FObject obj, XMLStreamWriter xmlStreamWriter) throws XMLStreamException {
+    try {
+      xmlStreamWriter.writeStartElement("object");
+      xmlStreamWriter.writeAttribute("class", obj.getClass().toString().replaceAll("class ", ""));
+      writeObjectProperties(obj, xmlStreamWriter);
+      xmlStreamWriter.writeEndElement();
+    } catch (XMLStreamException ex) {
+      throw new XMLStreamException("Error while writing xml file");
+    }
+  }
+
+  // Write properties from given FObject
+  public static void writeObjectProperties(FObject obj, XMLStreamWriter writer) {
+    List props = obj.getClassInfo().getAxioms();
+    Iterator propItr = props.iterator();
+
+    try {
+      while ( propItr.hasNext() ) {
+        PropertyInfo currentProp = (PropertyInfo) propItr.next();
+        Object value = currentProp.f(obj);
+        if ( value != null && value != "" ) {
+          writer.writeStartElement(currentProp.getName());
+          // Case for date
+          if ( value instanceof java.util.Date ) {
+            String dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'";
+            LocalDateTime ldt = LocalDateTime.ofInstant(((Date) value).toInstant(), ZoneId.of("UTC"));
+            DateTimeFormatter formatter = DateTimeFormatter.ofPattern(dateFormat).withZone(java.time.ZoneOffset.UTC);
+            String s = formatter.format(ldt);
+            writer.writeCharacters(s);
+          } else {
+            writer.writeCharacters(value.toString());
+          }
+          writer.writeEndElement();
+        }
+      }
+    } catch (XMLStreamException ex) {
+    }
+  }
+
+  // Returns XML string as partial XML string with only object tags
+  public static String toXMLString(FObject obj) {
+    XMLOutputFactory factory = XMLOutputFactory.newInstance();
+    StringWriter sw = new StringWriter();
+    try {
+      XMLStreamWriter writer = factory.createXMLStreamWriter(sw);
+      writer = new IndentingXMLStreamWriter(writer);
+      toXML(obj, writer);
+    } catch (XMLStreamException ex) {
+    }
+    return sw.toString();
+  }
+
+  // Returns XML string as full XML document string with document tags
+  public static String toXMLString(List<FObject> objArray) {
```

Comment:
This is going to copy each string once for each level of the XML, which is very very inefficient. Instead, pass the StringWriter down rather than passing Strings up.